tests/integration/service_process.py

```diff
@@ -0,0 +1,316 @@
+        self._stop_event.set()
+        if self._stdout_thread and self._stdout_thread.is_alive():
+            self._stdout_thread.join(timeout=0.1)
+            if self._stdout_thread.is_alive():
+                logger.warning("stdout thread did not stop for %s", self.service_module)
+        if self._stderr_thread and self._stderr_thread.is_alive():
+            self._stderr_thread.join(timeout=0.1)
+            if self._stderr_thread.is_alive():
+                logger.warning("stderr thread did not stop for %s", self.service_module)
+
+        # Close pipes
+        if self.process.stdout:
+            self.process.stdout.close()
+        if self.process.stderr:
+            self.process.stderr.close()
+
+    def is_running(self) -> bool:
+        """Check if the service process is currently running."""
+        if self.process is None:
+            return False
+        return self.process.poll() is None
+
+    def get_stdout(self) -> str:
+        """Get accumulated stdout from the service."""
+        return ''.join(self._stdout_lines)
+
+    def get_stderr(self) -> str:
+        """Get accumulated stderr from the service."""
+        return ''.join(self._stderr_lines)
+
+    def __enter__(self) -> 'ServiceProcess':
+        """Enter context manager - starts the service."""
+        self.start()
+        return self
+
+    def __exit__(
+        self,
+        exc_type: type[BaseException] | None,
+        exc_val: BaseException | None,
+        exc_tb: TracebackType | None,
+    ) -> None:
+        """Exit context manager - stops the service."""
+        self.stop()
+
+
+class ServiceGroup:
+    """
+    Manages multiple services as a group for integration testing.
+
+    This is useful for starting multiple interdependent services
+    (e.g., fake_monitors + monitor_data) with proper ordering and cleanup.
+
+    Parameters
+    ----------
+    services:
+        Dictionary mapping service names to ServiceProcess instances
+    """
+
+    def __init__(self, services: dict[str, ServiceProcess]):
+        self.services = services
+
+    def start_all(self, startup_delay: float = 2.0) -> None:
+        """
+        Start all services in the group.
+
+        Parameters
+        ----------
+        startup_delay:
+            Time to wait after starting each service (seconds)
+        """
+        for name, service in self.services.items():
+            logger.info("Starting service group member: %s", name)
+            service.start(startup_delay=startup_delay)
+
+    def stop_all(self, timeout: float = 10.0) -> None:
+        """
+        Stop all services in the group (in reverse order).
+
+        Parameters
+        ----------
+        timeout:
+            Maximum time to wait for each service to stop (seconds)
+        """
+        # Stop in reverse order
+        for name, service in reversed(list(self.services.items())):
+            logger.info("Stopping service group member: %s", name)
+            service.stop(timeout=timeout)
```

Comment:
If any service raises during stop or start, other services might not get stopped. I think you should use `ExitStack` or something similar here.